c_src/mongoose_jid.c

```diff
@@ -0,0 +1,73 @@
+#include "erl_nif.h"
+#include <stdio.h>
+#include <string.h>
+
+ERL_NIF_TERM
+mk_error(ErlNifEnv* env)
+{
+    ERL_NIF_TERM ret;
+    enif_make_existing_atom(env, "error", &ret, ERL_NIF_LATIN1);
+    return ret;
+}
+
+static ERL_NIF_TERM
+from_binary_nif(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
+{
+    if(argc != 1) {
+        return enif_make_badarg(env);
+    }
+
+    const ERL_NIF_TERM arg = argv[0];
+    ErlNifBinary bin;
+    if (!enif_inspect_binary(env, arg, &bin)) {
+        return enif_make_badarg(env);
+    }
+
+    const size_t size = bin.size;
+    unsigned commercial_at = -1;
+    unsigned slash = size;
+    for (unsigned i = 0; i < size ; ++i) {
+        switch(bin.data[i]) {
+            case '/':
+                if (slash == size) {
+                    slash = i;
+                    goto end_loop;
```

Comment:
I can't just approve a `goto` without a comment... I guess you did some profiling there, so let it be like it is.

Reply:
This was like that already when I wrote the original NIF over two years ago. I remember manually reading the assembly code generated and it was more optimal, but, honestly I can't recall any more details, would need to verify that again. Here the only diff is that as I'm not using any C++ construct really, I migrated the compilation to pure C99 instead of C++11.